model/stats/db.go

```diff
@@ -442,15 +443,309 @@ func statsToUpdatePipeline(projectId string, start time.Time, end time.Time) []b
+			{Name: "distro", Value: 1},
+		}},
+		{"$limit": filter.Limit},
+	}
+}
+
+// Builds the match stage of the test query pipeline based on the filter options.
+func buildMatchStageForTest(filter *StatsFilter) bson.M {
+	match := bson.M{
+		"_id.date": bson.M{
+			"$gte": filter.AfterDate,
+			"$lt":  filter.BeforeDate,
+		},
+		"_id.project":   filter.Project,
+		"_id.requester": bson.M{"$in": filter.Requesters},
+	}
+	addMatchIn(match, "_id.test_file", filter.Tests)
+	addMatchIn(match, "_id.task_name", filter.Tasks)
+	addMatchIn(match, "_id.variant", filter.BuildVariants)
+	addMatchIn(match, "_id.distro", filter.Distros)
+
+	if filter.StartAt != nil {
+		addMatchTestPagination(match, filter)
+	}
+
+	return bson.M{"$match": match}
+}
+
+// Builds the $addFields stage that sets the start date of the grouped period the stats document belongs in.
+func buildAddFieldsDateStage(fieldName string, start time.Time, end time.Time, numDays int) bson.M {
+	if numDays <= 1 {
+		return bson.M{"$addFields": bson.M{fieldName: "$_id.date"}}
+	}
+	boundaries := dateBoundaries(start, end, numDays)
+	branches := make([]bson.M, len(boundaries))
+	for i := 0; i < len(boundaries)-1; i++ {
+		branches[i] = bson.M{
+			"case": bson.M{"$and": array{
+				bson.M{"$gte": array{"$_id.date", boundaries[i]}},
+				bson.M{"$lt": array{"$_id.date", boundaries[i+1]}},
+			}},
+			"then": boundaries[i],
+		}
+	}
+	lastIndex := len(boundaries) - 1
+	branches[lastIndex] = bson.M{
+		"case": bson.M{"$gte": array{"$_id.date", boundaries[lastIndex]}},
+		"then": boundaries[lastIndex],
+	}
+	return bson.M{"$addFields": bson.M{fieldName: bson.M{"$switch": bson.M{"branches": branches}}}}
+}
+
+// Builds the _id field for the $group stage corresponding to the GroupBy value.
+func buildGroupId(groupBy GroupBy) bson.M {
+	id := bson.M{"date": "$date"}
+	switch groupBy {
+	case GroupByDistro:
+		id["distro"] = "$_id.distro"
+		fallthrough
+	case GroupByVariant:
+		id["variant"] = "$_id.variant"
+		fallthrough
+	case GroupByTask:
+		id["task_name"] = "$_id.task_name"
+		fallthrough
+	case GroupByTest:
+		id["test_file"] = "$_id.test_file"
+	}
+	return id
+}
+
+// Edits a match expression to include that the 'fieldName' field must have its value in 'values'.
+// Does nothing if 'values' is empty.
+func addMatchIn(matchExpr bson.M, fieldName string, values []string) {
+	if len(values) == 1 {
+		matchExpr[fieldName] = values[0]
+	} else if len(values) > 1 {
+		matchExpr[fieldName] = bson.M{"$in": values}
+	}
+}
+
+// Adds to an existing $match expression the conditions imposed by the filter StartAt field
+func addMatchTestPagination(matchExpr bson.M, filter *StatsFilter) {
+	var dateOperator string
+	if filter.Sort == SortEarliestFirst {
+		dateOperator = "$gt"
+	} else {
+		dateOperator = "$lt"
+	}
+
+	var fields []string
+	var operators []string
+	var values []interface{}
+
+	switch filter.GroupBy {
+	case GroupByTest:
+		fields = []string{"_id.date", "_id.test_file"}
```

Comment:
fields should probably use bsonutil.GetDottedFieldName() and string together from package variables.

Reply:
While I agree in principle, here the aggregations are already fairly complicated and I feel using variables will make the file a lot less readable. That's why I chose to keep the string literals.